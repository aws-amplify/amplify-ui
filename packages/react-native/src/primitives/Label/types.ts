import { TextProps, TextStyle } from 'react-native';

export interface LabelProps extends TextProps {
  /**
   * @description
   * The variant property will affect the color of the Label.
   */
  variant?: LabelVariant;
}

export type LabelPosition = 'start' | 'end' | 'top' | 'bottom';

export type LabelVariant =
  | 'primary'
  | 'secondary'
  | 'tertiary'
  | 'error'
  | 'warning'
  | 'info'
  | 'success';

export interface LabelStyles {
  text?: TextStyle;
  primary?: TextStyle;
  secondary?: TextStyle;
  tertiary?: TextStyle;
  error?: TextStyle;
  warning?: TextStyle;
  success?: TextStyle;
  info?: TextStyle;
}
