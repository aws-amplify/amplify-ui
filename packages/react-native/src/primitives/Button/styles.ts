import { StyleSheet } from 'react-native';

import { StrictTheme } from '../../theme';
import { ButtonStyles } from './types';

export const getThemedStyles = (theme: StrictTheme): ButtonStyles => {
  const { tokens, components } = theme;
  const {
    colors,
    fontSizes,
    fontWeights,
    opacities,
    space,
    radii,
    borderWidths,
  } = tokens;

  return StyleSheet.create({
    container: {
      alignItems: 'center',
      backgroundColor: colors.transparent,
      justifyContent: 'center',
      paddingVertical: space.small,
      paddingHorizontal: space.medium,
      borderRadius: radii.small,
      ...components?.button?.container,
    },
    containerDefault: {
      borderWidth: borderWidths.small,
      borderColor: colors.border.primary,
      ...components?.button?.containerDefault,
    },
    containerPrimary: {
      backgroundColor: colors.brand.primary[80],
      ...components?.button?.containerPrimary,
    },
    containerLink: {
      ...components?.button?.containerLink,
    },
    disabled: {
      opacity: opacities[60],
      ...components?.button?.disabled,
    },
    pressed: {
      opacity: opacities[60],
      ...components?.button?.pressed,
    },
    text: {
      textAlign: 'center',
      color: colors.font.primary,
      fontSize: fontSizes.medium,
      fontWeight: fontWeights.bold,
      ...components?.button?.text,
    },
    textPrimary: {
      color: colors.font.inverse,
      ...components?.button?.textPrimary,
    },
    textLink: {
      color: colors.font.interactive,
      ...components?.button?.textLink,
    },
  });
};
