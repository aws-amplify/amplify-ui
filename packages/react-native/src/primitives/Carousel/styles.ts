import { StyleSheet } from 'react-native';

import { CarouselStyles } from './types';

export const styles: CarouselStyles = StyleSheet.create({
  indicator: {
    bottom: 0,
    flexDirection: 'row',
    justifyContent: 'center',
    left: 0,
    position: 'absolute',
    right: 0,
  },
});
