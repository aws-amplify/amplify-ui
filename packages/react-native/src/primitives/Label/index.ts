export { default as Label } from './Label';
export { LabelProps, LabelStyles } from './types';
