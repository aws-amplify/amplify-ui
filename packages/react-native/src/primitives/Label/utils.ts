import { ViewStyle } from 'react-native';
import { LabelPosition } from './types';

const FLEX_DIRECTIONS: Record<LabelPosition, ViewStyle['flexDirection']> = {
  start: 'row-reverse',
  end: 'row',
  top: 'column-reverse',
  bottom: 'column',
};

/**
 * The `flexDirection` value returned from `getFlexDirectionFromLabelPosition`
 * assumes a `Label` component follows the `Primary` component it labels inside
 * the `Container` that the `flexDirection` is applied to.
 *
 * Example Usage:
 * ```jsx
 *   const flexDirection = getFlexDirectionFromLabelPosition(labelPosition);
 *   <Container style={{ flexDirection }}>
 *     <Primary />
 *     <Label />
 *   </Container>
 * ```
 */
export const getFlexDirectionFromLabelPosition = (
  position: LabelPosition
): ViewStyle['flexDirection'] => FLEX_DIRECTIONS[position];
