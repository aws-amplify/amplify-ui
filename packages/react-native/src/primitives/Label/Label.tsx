import React from 'react';
import { Text } from 'react-native';

import { useTheme } from '../../theme';
import { LabelProps } from './types';
import { getThemedStyles } from './styles';

export default function Label({
  accessibilityRole = 'text',
  children,
  style,
  variant = 'primary',
  ...rest
}: LabelProps): JSX.Element {
  const theme = useTheme();
  const themedStyle = getThemedStyles(theme);
  const textStyle = [themedStyle.text, themedStyle[variant], style];

  return (
    <Text {...rest} accessibilityRole={accessibilityRole} style={textStyle}>
      {children}
    </Text>
  );
}
