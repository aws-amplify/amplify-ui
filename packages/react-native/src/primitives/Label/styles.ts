import { StyleSheet } from 'react-native';

import { StrictTheme } from '../../theme';
import { getLineHeight } from '../../utils';
import { LabelStyles } from './types';

export const getThemedStyles = (theme: StrictTheme): LabelStyles => {
  const {
    components,
    tokens: { colors, fontSizes, fontWeights },
  } = theme;

  return StyleSheet.create({
    text: {
      fontSize: fontSizes.medium,
      fontWeight: fontWeights.normal,
      lineHeight: getLineHeight(fontSizes.medium),
      ...components?.label?.text,
    },
    primary: {
      color: colors.font.primary,
      ...components?.label?.primary,
    },
    secondary: {
      color: colors.font.secondary,
      ...components?.label?.secondary,
    },
    tertiary: {
      color: colors.font.tertiary,
      ...components?.label?.tertiary,
    },
    error: {
      color: colors.font.error,
      ...components?.label?.error,
    },
    warning: {
      color: colors.font.warning,
      ...components?.label?.warning,
    },
    success: {
      color: colors.font.success,
      ...components?.label?.success,
    },
    info: {
      color: colors.font.info,
      ...components?.label?.info,
    },
  });
};
