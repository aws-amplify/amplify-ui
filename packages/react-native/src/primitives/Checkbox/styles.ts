import { StyleSheet } from 'react-native';

import { StrictTheme } from '../../theme';
import { CheckboxStyles } from './types';
import { getFlexDirectionFromLabelPosition } from '../Label/utils';
import { LabelPosition } from '../Label/types';

export const getThemedStyles = (
  theme: StrictTheme,
  labelPosition: LabelPosition
): CheckboxStyles => {
  const { tokens, components } = theme;
  const { opacities, space } = tokens;

  return StyleSheet.create({
    container: {
      alignItems: 'center',
      padding: space.xs,
      flexDirection: getFlexDirectionFromLabelPosition(labelPosition),
      ...components?.checkbox?.container,
    },
    disabled: {
      opacity: opacities[60],
      ...components?.checkbox?.disabled,
    },
    icon: {
      ...components?.checkbox?.icon,
    },
    label: {
      ...components?.checkbox?.label,
    },
  });
};
