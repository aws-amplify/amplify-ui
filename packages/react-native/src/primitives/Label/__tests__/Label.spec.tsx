import React from 'react';
import { render, renderHook } from '@testing-library/react-native';

import { useTheme } from '../../../theme';
import { getThemedStyles } from '../styles';
import Label from '../Label';

const text = 'Default Label';
const testID = 'labelID';
const props = {
  testID,
};

describe('Label', () => {
  it('renders a default Label', () => {
    const { toJSON, getByText, getByRole } = render(
      <Label {...props}>{text}</Label>
    );
    expect(toJSON()).toMatchSnapshot();

    expect(getByText(text)).toBeDefined();
    expect(getByRole('text')).toBeDefined();
  });

  it('applies accessibility role', () => {
    const { toJSON, queryByRole, getByRole } = render(
      <Label {...props} accessibilityRole="none">
        {text}
      </Label>
    );
    expect(toJSON()).toMatchSnapshot();

    expect(queryByRole('text')).toBe(null);
    expect(getByRole('none')).toBeDefined();
  });

  it('should apply theme and style props', () => {
    const customStyle = { color: 'red' };

    const { toJSON, getByTestId } = render(
      <Label {...props} style={customStyle}>
        Red Label
      </Label>
    );

    const { result } = renderHook(() => useTheme());
    const themedStyle = getThemedStyles(result.current);

    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId(testID).props.style).toStrictEqual([
      themedStyle.text,
      themedStyle['primary'],
      customStyle,
    ]);
  });

  it('should apply variant styles', () => {
    const variant = 'success';

    const { toJSON, getByTestId } = render(
      <Label {...props} variant={variant}>
        Label
      </Label>
    );
    const { result } = renderHook(() => useTheme());
    const themedStyle = getThemedStyles(result.current);

    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId(testID).props.style).toStrictEqual([
      themedStyle.text,
      themedStyle[variant],
      undefined, // no custom styles
    ]);
  });
});
