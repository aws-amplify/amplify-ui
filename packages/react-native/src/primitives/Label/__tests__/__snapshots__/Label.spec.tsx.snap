// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label applies accessibility role 1`] = `
<Text
  accessibilityRole="none"
  style={
    Array [
      Object {
        "fontSize": 16,
        "fontWeight": "400",
        "marginHorizontal": 4,
        "marginVertical": 2,
      },
      Object {
        "color": "hsl(210, 50%, 10%)",
      },
      undefined,
    ]
  }
  testID="labelID"
>
  Default Label
</Text>
`;

exports[`Label renders a default Label 1`] = `
<Text
  accessibilityRole="text"
  style={
    Array [
      Object {
        "fontSize": 16,
        "fontWeight": "400",
        "marginHorizontal": 4,
        "marginVertical": 2,
      },
      Object {
        "color": "hsl(210, 50%, 10%)",
      },
      undefined,
    ]
  }
  testID="labelID"
>
  Default Label
</Text>
`;

exports[`Label should apply theme and style props 1`] = `
<Text
  accessibilityRole="text"
  style={
    Array [
      Object {
        "fontSize": 16,
        "fontWeight": "400",
        "marginHorizontal": 4,
        "marginVertical": 2,
      },
      Object {
        "color": "hsl(210, 50%, 10%)",
      },
      Object {
        "color": "red",
      },
    ]
  }
  testID="labelID"
>
  Red Label
</Text>
`;

exports[`Label should apply variant styles 1`] = `
<Text
  accessibilityRole="text"
  style={
    Array [
      Object {
        "fontSize": 16,
        "fontWeight": "400",
        "marginHorizontal": 4,
        "marginVertical": 2,
      },
      Object {
        "color": "hsl(130, 27%, 29%)",
      },
      undefined,
    ]
  }
  testID="labelID"
>
  Label
</Text>
`;
